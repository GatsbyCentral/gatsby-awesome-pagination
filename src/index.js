// @flow

import isString from "lodash/fp/isString";
import get from "lodash/fp/get";
import times from "lodash/fp/times";
import cloneDeep from "lodash/fp/cloneDeep";
import isInteger from "lodash/fp/isInteger";

import {
  paginatedPath,
  getPreviousItem,
  getNextItem,
  calculateSkip
} from "./utils";

import type { PathPrefix } from './utils'

type CreatePage = ({}) => void;

type PaginateOpts = {
  createPage: CreatePage,
  items: {}[],
  itemsPerPage: number,
  itemsPerFirstPage?: number,
  pathPrefix: PathPrefix,
  component: string,
  context?: {}
};
export const paginate = (opts: PaginateOpts): void => {
  const {
    createPage,
    items,
    itemsPerPage,
    itemsPerFirstPage,
    pathPrefix,
    component,
    context
  } = opts;

  // How many items do we have in total? We use `items.length` here. In fact, we
  // could just accept an integer in the API as the actual contents of `items`
  // is never used.
  const totalItems = items.length;
  // If `itemsPerFirstPage` is specified, use that value for the first page,
  // otherwise use `itemsPerPage`.
  // $FlowExpectError
  const firstPageCount: number = isInteger(itemsPerFirstPage)
    ? itemsPerFirstPage
    : itemsPerPage;

  // How many page should we have?
  const numberOfPages =
    // If there are less than `firstPageCount` items, we'll only have 1 page
    totalItems <= firstPageCount
      ? 1
      : Math.ceil((totalItems - firstPageCount) / itemsPerPage) + 1;

  // Iterate as many times as we need pages
  times((pageNumber: number) => {
    // Create the path for this page
    const path = paginatedPath(pathPrefix, pageNumber, numberOfPages);

    // Calculate the path for the previous and next pages
    const previousPagePath = paginatedPath(
      pathPrefix,
      pageNumber - 1,
      numberOfPages
    );
    const nextPagePath = paginatedPath(
      pathPrefix,
      pageNumber + 1,
      numberOfPages
    );

    // Call `createPage()` for this paginated page
    createPage({
      path,
      component,
      // Clone the passed `context` and extend our new pagination context values
      // on top of it.
      context: Object.assign({}, cloneDeep(context), {
        pageNumber,
        humanPageNumber: pageNumber + 1,
        skip: calculateSkip(pageNumber, firstPageCount, itemsPerPage),
        limit: pageNumber === 0 ? firstPageCount : itemsPerPage,
        numberOfPages,
        previousPagePath,
        nextPagePath
      })
    });
  })(numberOfPages);
};

type CreatePagePerItemOpts = {
  createPage: CreatePage,
  items: {}[],
  itemToPath: string | (({}) => string),
  itemToId: string | (({}) => string),
  component: string
};
export const createPagePerItem = (opts: CreatePagePerItemOpts): void => {
  const { createPage, items, itemToPath, itemToId, component } = opts;

  // $FlowExpectError
  const getPath: ({}) => string = isString(itemToPath)
    ? get(itemToPath)
    : itemToPath;
  // $FlowExpectError
  const getId: ({}) => string = isString(itemToId) ? get(itemToId) : itemToId;

  // We cannot use `forEach()` here because in the FP version of lodash, the
  // iteratee is capped to a single argument, the item itself. We cannot get the
  // item's index in the array. So instead we use `times()` and provide the
  // length of the array.
  times((index: number) => {
    const item = items[index];
    const path = getPath(item);
    const id = getId(item);

    // NOTE: If there is no previous / next item, we set an empty string as the
    // value for the next and previous path and ID. Gatsby ignores context
    // values which are undefined, so we need these to exist.
    const previousItem = getPreviousItem(items, index);
    const previousPath = getPath(previousItem) || "";
    const nextItem = getNextItem(items, index);
    const nextPath = getPath(nextItem) || "";

    // Does the item have a `context` field?
    const itemContext = get("context")(item) || {};
    const context = Object.assign({}, itemContext, {
      pageId: id,
      previousPagePath: previousPath,
      previousItem: previousItem,
      nextPagePath: nextPath,
      nextItem: nextItem
    });

    // Call `createPage()` for this item
    createPage({
      path,
      component,
      context
    });
  })(items.length);
};
