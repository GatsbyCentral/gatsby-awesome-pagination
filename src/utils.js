// @flow
import get from "lodash/fp/get";

type ReturnedItem = {} | void;

export const getPreviousItem = (items: {}[], index: number): ReturnedItem => {
  // If this is the first (or -1) element, there is no previous, so return
  // undefined
  if (index <= 0) {
    return;
  }

  // Get the previous element
  return get(`[${index - 1}]`, items);
};

export const getNextItem = (items: {}[], index: number): ReturnedItem => {
  // If this is the last element (or later), there is no previous, so return
  // undefined
  if (index >= items.length - 1) {
    return;
  }

  // Get the previous element
  return get(`[${index + 1}]`, items);
};

export type PathPrefix = string | ({
  pageNumber: number,
  numberOfPages: number,
}) => string;

export const paginatedPath = (
  pathPrefix: PathPrefix,
  pageNumber: number,
  numberOfPages: number,
): string => {
  // If this page is less than zero (-1 for example), then it  it does not
  // exist, return an empty string.
  if (pageNumber < 0) {
    return "";
  }

  // If this page number (which is zero indexed) plus one is more than the total
  // number of pages, then this page does not exist, so return an empty string.
  if (pageNumber + 1 > numberOfPages) {
    return "";
  }

  const computedPathPrefix = typeof pathPrefix === "function"
    ? pathPrefix({ pageNumber, numberOfPages })
    : pathPrefix;

  // If this is page 0, return only the pathPrefix
  if (pageNumber === 0) {
    return computedPathPrefix;
  }

  // Othewrise, add a slash and the number + 1. We add 1 because `pageNumber` is
  // zero indexed, but for human consuption, we want 1 indexed numbers.
  return `${computedPathPrefix !== "/" ? computedPathPrefix : ""}/${pageNumber + 1}`;
  // NOTE: If `pathPrefix` is a single slash (the index page) then we do not
  // want to output two slashes, so we omit it.
};

export const calculateSkip = (
  pageNumber: number,
  firstPageCount: number,
  itemsPerPage: number
): number => {
  if (pageNumber === 0) {
    return 0;
  } else if (pageNumber === 1) {
    return firstPageCount;
  } else {
    return firstPageCount + itemsPerPage * (pageNumber - 1);
  }
};
