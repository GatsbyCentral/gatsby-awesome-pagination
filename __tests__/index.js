import { paginate, createPagePerItem } from "../src"

describe("paginate", () => {
  it("creates pages and forwards context", () => {
    const createPage = jest.fn();
    paginate({
      createPage,
      items: Array(12).fill(),
      itemsPerPage: 3,
      itemsPerFirstPage: 5,
      pathPrefix: "/blog",
      component: "path/to/blog",
      context: { additionalContext: true }
    });
    expect(createPage).toMatchSnapshot();
  });

  it("allows fine-tuning pathPrefix with a function", () => {
    const createPage = jest.fn();
    paginate({
      createPage,
      items: Array(12).fill(),
      itemsPerPage: 3,
      pathPrefix: ({ pageNumber }) => pageNumber === 0 ? "/blog" : "/blog/page",
      component: "path/to/blog"
    });
    expect(createPage).toMatchSnapshot();
  });
});
