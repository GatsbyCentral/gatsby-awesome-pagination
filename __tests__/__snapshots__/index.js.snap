// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`paginate allows fine-tuning pathPrefix with a function 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "component": "path/to/blog",
        "context": Object {
          "humanPageNumber": 1,
          "limit": 3,
          "nextPagePath": "/blog/page/2",
          "numberOfPages": 4,
          "pageNumber": 0,
          "previousPagePath": "",
          "skip": 0,
        },
        "path": "/blog",
      },
    ],
    Array [
      Object {
        "component": "path/to/blog",
        "context": Object {
          "humanPageNumber": 2,
          "limit": 3,
          "nextPagePath": "/blog/page/3",
          "numberOfPages": 4,
          "pageNumber": 1,
          "previousPagePath": "/blog",
          "skip": 3,
        },
        "path": "/blog/page/2",
      },
    ],
    Array [
      Object {
        "component": "path/to/blog",
        "context": Object {
          "humanPageNumber": 3,
          "limit": 3,
          "nextPagePath": "/blog/page/4",
          "numberOfPages": 4,
          "pageNumber": 2,
          "previousPagePath": "/blog/page/2",
          "skip": 6,
        },
        "path": "/blog/page/3",
      },
    ],
    Array [
      Object {
        "component": "path/to/blog",
        "context": Object {
          "humanPageNumber": 4,
          "limit": 3,
          "nextPagePath": "",
          "numberOfPages": 4,
          "pageNumber": 3,
          "previousPagePath": "/blog/page/3",
          "skip": 9,
        },
        "path": "/blog/page/4",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`paginate creates pages and forwards context 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "component": "path/to/blog",
        "context": Object {
          "additionalContext": true,
          "humanPageNumber": 1,
          "limit": 5,
          "nextPagePath": "/blog/2",
          "numberOfPages": 4,
          "pageNumber": 0,
          "previousPagePath": "",
          "skip": 0,
        },
        "path": "/blog",
      },
    ],
    Array [
      Object {
        "component": "path/to/blog",
        "context": Object {
          "additionalContext": true,
          "humanPageNumber": 2,
          "limit": 3,
          "nextPagePath": "/blog/3",
          "numberOfPages": 4,
          "pageNumber": 1,
          "previousPagePath": "/blog",
          "skip": 5,
        },
        "path": "/blog/2",
      },
    ],
    Array [
      Object {
        "component": "path/to/blog",
        "context": Object {
          "additionalContext": true,
          "humanPageNumber": 3,
          "limit": 3,
          "nextPagePath": "/blog/4",
          "numberOfPages": 4,
          "pageNumber": 2,
          "previousPagePath": "/blog/2",
          "skip": 8,
        },
        "path": "/blog/3",
      },
    ],
    Array [
      Object {
        "component": "path/to/blog",
        "context": Object {
          "additionalContext": true,
          "humanPageNumber": 4,
          "limit": 3,
          "nextPagePath": "",
          "numberOfPages": 4,
          "pageNumber": 3,
          "previousPagePath": "/blog/3",
          "skip": 11,
        },
        "path": "/blog/4",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
